src/bencode_parser.cpp

```diff
@@ -0,0 +1,191 @@
+#include "bencode_parser.hpp"
+
+#include <regex>
+
+#include "memory"
+
+using namespace fur::bencode;
+
+std::string BencodeParser::encode(const BencodeValue& value) {
+  return value.to_string();
+}
+
+std::vector<std::string> BencodeParser::tokenizer(const std::string& encoded) {
+  std::vector<std::string> tokens;
+  std::regex regexp("([idel])|(\\d+):|(-?\\d+)");
+  std::sregex_token_iterator iter(encoded.begin(), encoded.end(), regexp,
+                                  {-1, 0});
+  std::sregex_token_iterator end;
+  while (iter != end) {
+    if (iter->length() > 0) {
+      tokens.push_back(*iter);
+    }
+    ++iter;
+  }
+  return tokens;
+}
+
+std::unique_ptr<BencodeValue> BencodeParser::decode(
+    const std::string& decoded) {
+  std::vector<std::string> tokens = tokenizer(decoded);
+  if (tokens.empty()) {
+    throw std::invalid_argument(
+        "BencodeParser::decode(std::string decoded): invalid decoded string");
+  }
+  _tokens = tokens;
+  _index = 0;  // Reset the index
+  return decode();
+}
+
+// Given a vector of tokens, returns a BencodeValue object
+std::unique_ptr<BencodeValue> BencodeParser::decode() {
+  std::regex reg_string("^\\d+:$");
+  if (!_tokens.empty()) {
+    auto token = _tokens[_index];
+    if (token == "i") {
+      // If the first token is "i", then it's an integer, so we have a fixed
+      // length
+      if (_index == 0 && _tokens.size() != 3) {
```

Comment:
I think this is supposed to catch something like `i1e i2e e' (missing the `l` to begin a list at the very beginning) but it would still miss something like `l i1e i2e e e`.